PR: https://github.com/odoo/odoo/pull/173288

From: 75b385bae036031b8fef651977bad71691b1e7a2
From: Yannick Tivisse
Date: 2024-08-26 10:55:02

Breaking data model changes scores: del:2.4000000000000004 + add:2, change matches:
-    applicant_skill_ids = fields.One2many('hr.applicant.skill', 'applicant_id', string="Skills")
-    is_interviewer = fields.Boolean(compute='_compute_is_interviewer')
+    candidate_skill_ids = fields.One2many(related="candidate_id.candidate_skill_ids", readonly=False)
+    skill_ids = fields.Many2many(related="candidate_id.skill_ids", readonly=False)
+    candidate_skill_ids = fields.One2many('hr.candidate.skill', 'candidate_id', string="Skills")
-    applicant_id = fields.Many2one(

Total Changes: 294

[IMP] hr_recruitment: Introduce hr.candidate model

Purpose
=======

Based on this task https://www.odoo.com/odoo/9285/tasks/4037988/9285/project.task/4050307?debug=1&cids=1
and many feedbacks we got before (leading partially to Headhunter industry), we realize that if we
want to cover the whole market of ATS we need a strong solution.

Specification
=============

We split hr.applicant between candidates and applications

A candidate is created outside of a job position, it contains all data that will
never change (skills, name, contact info, expected salary), it gathers history
(applied job positions, states, ...)

An application is the link between an candidate and a Job Position, it contains all related
data (source, medium, proposed salary, evaluation, ...) and it got a state
(first interview, contract proposed, refused)

In a Job position, you would see distinct applicants, but through the hr.applications model.
This mechanism is similar to a vehicle which is the link between a Model and a Driver.

The list view "all applications" would list "All Applicants" and each Applicant would have a
smartbutton "Applications"

closes odoo/odoo#173288

Details: https://link.excalidraw.com/l/65VNwvy7c4X/5SjjwEHMIPG
Taskid: 4050442
Related: odoo/enterprise#66734
Related: odoo/upgrade#6300
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_recruitment_skills/models/__init__.py
+++ b/addons/hr_recruitment_skills/models/__init__.py
@@ -2,5 +2,6 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from . import hr_applicant
-from . import hr_applicant_skill
+from . import hr_candidate
+from . import hr_candidate_skill
 from . import hr_job

--- a/addons/hr_recruitment_skills/models/hr_applicant.py
+++ b/addons/hr_recruitment_skills/models/hr_applicant.py
@@ -1,75 +1,10 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from ast import literal_eval
-
-from odoo import fields, models, api
+from odoo import fields, models
 
 
 class HrApplicant(models.Model):
     _inherit = 'hr.applicant'
 
-    applicant_skill_ids = fields.One2many('hr.applicant.skill', 'applicant_id', string="Skills")
-    skill_ids = fields.Many2many('hr.skill', compute='_compute_skill_ids', store=True)
-    is_interviewer = fields.Boolean(compute='_compute_is_interviewer')
-    matching_skill_ids = fields.Many2many(comodel_name='hr.skill', string="Matching Skills", compute="_compute_matching_skill_ids")
-    missing_skill_ids = fields.Many2many(comodel_name='hr.skill', string="Missing Skills", compute="_compute_matching_skill_ids")
-    matching_score = fields.Float(string="Matching Score(%)", compute="_compute_matching_skill_ids")
-
-    @api.depends_context('uid')
-    @api.depends('interviewer_ids', 'job_id.interviewer_ids')
-    def _compute_is_interviewer(self):
-        is_recruiter = self.env.user.has_group('hr_recruitment.group_hr_recruitment_user')
-        for applicant in self:
-            applicant.is_interviewer = not is_recruiter and self.env.user in (applicant.interviewer_ids | applicant.job_id.interviewer_ids)
-
-    @api.depends('applicant_skill_ids.skill_id')
-    def _compute_skill_ids(self):
-        for applicant in self:
-            applicant.skill_ids = applicant.applicant_skill_ids.skill_id
-
-    @api.depends_context('active_id')
-    @api.depends('skill_ids')
-    def _compute_matching_skill_ids(self):
-        job_id = self.env.context.get('active_id')
-        if not job_id:
-            self.matching_skill_ids = False
-            self.missing_skill_ids = False
-            self.matching_score = 0
-        else:
-            for applicant in self:
-                job_skills = self.env['hr.job'].browse(job_id).skill_ids
-                applicant.matching_skill_ids = job_skills & applicant.skill_ids
-                applicant.missing_skill_ids = job_skills - applicant.skill_ids
-                applicant.matching_score = (len(applicant.matching_skill_ids) / len(job_skills)) * 100 if job_skills else 0
-
-    def _get_employee_create_vals(self):
-        vals = super()._get_employee_create_vals()
-        vals['employee_skill_ids'] = [(0, 0, {
-            'skill_id': applicant_skill.skill_id.id,
-            'skill_level_id': applicant_skill.skill_level_id.id,
-            'skill_type_id': applicant_skill.skill_type_id.id,
-        }) for applicant_skill in self.applicant_skill_ids]
-        return vals
-
-    def _update_employee_from_applicant(self):
-        vals_list = []
-        for applicant in self:
-            existing_skills = applicant.emp_id.employee_skill_ids.skill_id
-            skills_to_create = applicant.applicant_skill_ids.skill_id - existing_skills
-            vals_list.extend([{
-                'employee_id': applicant.emp_id.id,
-                'skill_id': skill.id,
-                'skill_level_id': applicant.applicant_skill_ids.filtered(lambda s: s.skill_id == skill).skill_level_id.id,
-                'skill_type_id': skill.skill_type_id.id,
-            } for skill in skills_to_create])
-        self.env['hr.employee.skill'].create(vals_list)
-        return super()._update_employee_from_applicant()
-
-    def action_add_to_job(self):
-        self.with_context(just_moved=True).write({
-            'job_id': self.env['hr.job'].browse(self.env.context.get('active_id')).id,
-            'stage_id': self.env.ref('hr_recruitment.stage_job0'),
-        })
-        action = self.env['ir.actions.actions']._for_xml_id('hr_recruitment.action_hr_job_applications')
-        action['context'] = literal_eval(action['context'].replace('active_id', str(self.job_id.id)))
-        return action
+    candidate_skill_ids = fields.One2many(related="candidate_id.candidate_skill_ids", readonly=False)
+    skill_ids = fields.Many2many(related="candidate_id.skill_ids", readonly=False)

--- a/None
+++ b/addons/hr_recruitment_skills/models/hr_candidate.py
@@ -0,0 +1,68 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from ast import literal_eval
+
+from odoo import fields, models, api
+
+
+class HrCandidate(models.Model):
+    _inherit = 'hr.candidate'
+
+    candidate_skill_ids = fields.One2many('hr.candidate.skill', 'candidate_id', string="Skills")
+    skill_ids = fields.Many2many('hr.skill', compute='_compute_skill_ids', store=True)
+    matching_skill_ids = fields.Many2many(comodel_name='hr.skill', string="Matching Skills", compute="_compute_matching_skill_ids")
+    missing_skill_ids = fields.Many2many(comodel_name='hr.skill', string="Missing Skills", compute="_compute_matching_skill_ids")
+    matching_score = fields.Float(string="Matching Score(%)", compute="_compute_matching_skill_ids")
+
+    @api.depends_context('active_id')
+    @api.depends('skill_ids')
+    def _compute_matching_skill_ids(self):
+        job_id = self.env.context.get('active_id')
+        if not job_id:
+            self.matching_skill_ids = False
+            self.missing_skill_ids = False
+            self.matching_score = 0
+        else:
+            for candidate in self:
+                job_skills = self.env['hr.job'].browse(job_id).skill_ids
+                candidate.matching_skill_ids = job_skills & candidate.skill_ids
+                candidate.missing_skill_ids = job_skills - candidate.skill_ids
+                candidate.matching_score = (len(candidate.matching_skill_ids) / len(job_skills)) * 100 if job_skills else 0
+
+    @api.depends('candidate_skill_ids.skill_id')
+    def _compute_skill_ids(self):
+        for candidate in self:
+            candidate.skill_ids = candidate.candidate_skill_ids.skill_id
+
+    def _update_employee_from_candidate(self):
+        vals_list = []
+        for candidate in self:
+            existing_skills = candidate.employee_id.employee_skill_ids.skill_id
+            skills_to_create = candidate.candidate_skill_ids.skill_id - existing_skills
+            vals_list.extend([{
+                'employee_id': candidate.employee_id.id,
+                'skill_id': skill.id,
+                'skill_level_id': candidate.candidate_skill_ids.filtered(lambda s: s.skill_id == skill).skill_level_id.id,
+                'skill_type_id': skill.skill_type_id.id,
+            } for skill in skills_to_create])
+        self.env['hr.employee.skill'].create(vals_list)
+        return super()._update_employee_from_candidate()
+
+    def _get_employee_create_vals(self):
+        vals = super()._get_employee_create_vals()
+        vals['employee_skill_ids'] = [(0, 0, {
+            'skill_id': candidate_skill.skill_id.id,
+            'skill_level_id': candidate_skill.skill_level_id.id,
+            'skill_type_id': candidate_skill.skill_type_id.id,
+        }) for candidate_skill in self.candidate_skill_ids]
+        return vals
+
+    def action_create_application(self):
+        job = self.env['hr.job'].browse(self.env.context.get('active_id'))
+        self.env['hr.applicant'].with_context(just_moved=True).create([{
+            'candidate_id': candidate.id,
+            'job_id': job.id,
+        } for candidate in self])
+        action = self.env['ir.actions.actions']._for_xml_id('hr_recruitment.action_hr_job_applications')
+        action['context'] = literal_eval(action['context'].replace('active_id', str(job.id)))
+        return action

--- a/addons/hr_recruitment_skills/models/hr_applicant_skill.py
+++ b/addons/hr_recruitment_skills/models/hr_candidate_skill.py
@@ -1,18 +1,17 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import api, fields, models, _
 from odoo.exceptions import ValidationError
 
 
-class ApplicantSkill(models.Model):
-    _name = 'hr.applicant.skill'
-    _description = "Skill level for an applicant"
+class CandidateSkill(models.Model):
+    _name = 'hr.candidate.skill'
+    _description = "Skill level for a candidate"
     _rec_name = 'skill_id'
     _order = "skill_level_id"
 
-    applicant_id = fields.Many2one(
-        comodel_name='hr.applicant',
+    candidate_id = fields.Many2one(
+        comodel_name='hr.candidate',
         required=True,
         ondelete='cascade')
     skill_id = fields.Many2one(
@@ -36,32 +35,32 @@ class ApplicantSkill(models.Model):
         related='skill_level_id.level_progress')
 
     _sql_constraints = [
-        ('_unique_skill', 'unique (applicant_id, skill_id)', "Two levels for the same skill is not allowed"),
+        ('_unique_skill', 'unique (candidate_id, skill_id)', "Two levels for the same skill is not allowed"),
     ]
 
     @api.constrains('skill_id', 'skill_type_id')
     def _check_skill_type(self):
-        for applicant in self:
-            if applicant.skill_id not in applicant.skill_type_id.skill_ids:
-                raise ValidationError(_("The skill %(name)s and skill type %(type)s doesn't match", name=applicant.skill_id.name, type=applicant.skill_type_id.name))
+        for candidate_skill in self:
+            if candidate_skill.skill_id not in candidate_skill.skill_type_id.skill_ids:
+                raise ValidationError(_("The skill %(name)s and skill type %(type)s doesn't match", name=candidate_skill.skill_id.name, type=candidate_skill.skill_type_id.name))
 
     @api.constrains('skill_type_id', 'skill_level_id')
     def _check_skill_level(self):
-        for applicant in self:
-            if applicant.skill_level_id not in applicant.skill_type_id.skill_level_ids:
-                raise ValidationError(_("The skill level %(level)s is not valid for skill type: %(type)s", level=applicant.skill_level_id.name, type=applicant.skill_type_id.name))
+        for candidate_skill in self:
+            if candidate_skill.skill_level_id not in candidate_skill.skill_type_id.skill_level_ids:
+                raise ValidationError(_("The skill level %(level)s is not valid for skill type: %(type)s", level=candidate_skill.skill_level_id.name, type=candidate_skill.skill_type_id.name))
 
     @api.depends('skill_type_id')
     def _compute_skill_id(self):
-        for applicant in self:
-            if applicant.skill_id.skill_type_id != applicant.skill_type_id:
-                applicant.skill_id = False
+        for candidate_skill in self:
+            if candidate_skill.skill_id.skill_type_id != candidate_skill.skill_type_id:
+                candidate_skill.skill_id = False
 
     @api.depends('skill_id')
     def _compute_skill_level_id(self):
-        for applicant in self:
-            if not applicant.skill_id:
-                applicant.skill_level_id = False
+        for candidate_skill in self:
+            if not candidate_skill.skill_id:
+                candidate_skill.skill_level_id = False
             else:
-                skill_levels = applicant.skill_type_id.skill_level_ids
-                applicant.skill_level_id = skill_levels.filtered('default_level') or skill_levels[0] if skill_levels else False
+                skill_levels = candidate_skill.skill_type_id.skill_level_ids
+                candidate_skill.skill_level_id = skill_levels.filtered('default_level') or skill_levels[0] if skill_levels else False

--- a/addons/hr_recruitment_skills/models/hr_job.py
+++ b/addons/hr_recruitment_skills/models/hr_job.py
@@ -9,23 +9,22 @@ class HrJob(models.Model):
 
     skill_ids = fields.Many2many(comodel_name='hr.skill', string="Expected Skills")
 
-    def action_search_matching_applicant(self):
-        help_message_1 = _("No Matching Applicants")
-        help_message_2 = _("We do not have any applicants who meet the skill requirements for this job position in the database at the moment.")
-        action = self.env['ir.actions.actions']._for_xml_id('hr_recruitment.crm_case_categ0_act_job')
+    def action_search_matching_candidates(self):
+        self.ensure_one()
+        help_message_1 = _("No Matching Candidates")
+        help_message_2 = _("We do not have any candidates who meet the skill requirements for this job position in the database at the moment.")
+        action = self.env['ir.actions.actions']._for_xml_id('hr_recruitment.action_hr_candidate')
         context = literal_eval(action['context'])
         context['active_id'] = self.id
+        matching_candidates = self.env['hr.candidate'].search([('skill_ids', 'in', self.skill_ids.ids)]).filtered(lambda c: self.id not in c.applicant_ids.job_id.ids)
         action.update({
-            'name': _("Matching Applicants"),
+            'name': _("Matching Candidates"),
             'views': [
-                (self.env.ref('hr_recruitment_skills.crm_case_tree_view_inherit_hr_recruitment_skills').id, 'tree'),
+                (self.env.ref('hr_recruitment_skills.hr_candidate_view_tree').id, 'tree'),
                 (False, 'form'),
             ],
             'context': context,
-            'domain': [
-                ('job_id', '!=', self.id),
-                ('skill_ids', 'in', self.skill_ids.ids),
-            ],
+            'domain': [('id', 'in', matching_candidates.ids)],
             'help': Markup("<p class='o_view_nocontent_empty_folder'>%s</p><p>%s</p>") % (help_message_1, help_message_2),
         })
         return action
